snuba/api.py

```diff
@@ -344,55 +348,45 @@ def sdk_distribution(validated_body, timer):
 if application.debug or application.testing:
     # These should only be used for testing/debugging. Note that the database name
     # is checked to avoid scary production mishaps.
-    assert settings.CLICKHOUSE_TABLE in ('dev', 'test')
 
-    _ensured = False
+    _ensured = {}
 
-    def ensure_table_exists(force=False):
+    def ensure_table_exists(dataset, force=False, drop=False):
         global _ensured
 
-        if not force and _ensured:
+        if not force and _ensured.get(dataset, False):
             return
 
-        from snuba.clickhouse import get_table_definition, get_test_engine
+        # only run table creation statements on non-prod datasets
+        if not dataset.PROD:
+            if drop:
+                clickhouse_rw.execute(dataset.SCHEMA.get_local_table_drop())
+            clickhouse_rw.execute(dataset.SCHEMA.get_local_table_definition())
+            # TODO this is awkward, table creating above returns a
+            # statement to be executed here, but migrate takes a clickhouse
+            # connection to execute statements on
+            dataset.SCHEMA.migrate(clickhouse_rw)
 
-        clickhouse_rw.execute(
-            get_table_definition(
-                name=settings.CLICKHOUSE_TABLE,
-                engine=get_test_engine(),
-            )
-        )
-
-        if settings.CLICKHOUSE_TABLE == 'dev':
-            from snuba import migrate
-            migrate.run(clickhouse_rw, settings.CLICKHOUSE_TABLE)
-
-        _ensured = True
+        _ensured[dataset] = True
 
     @application.route('/tests/insert', methods=['POST'])
     def write():
-        from snuba.processor import process_message
-        from snuba.writer import row_from_processed_event, write_rows
+        from snuba.writer import write_rows
+        # TODO we need to make this work for multiple datasets
+        dataset = settings.get_dataset('events')
+        ensure_table_exists(dataset)
 
-        ensure_table_exists()
         body = json.loads(request.data)
-
-        rows = []
-        for event in body:
-            _, processed = process_message(event)
-            row = row_from_processed_event(processed)
-            rows.append(row)
-
-        write_rows(
-            clickhouse_rw,
-            table=settings.CLICKHOUSE_TABLE,
-            rows=rows
-        )
+        rows = [dataset.PROCESSOR.process_insert((2, 'insert', event)) for event in body]
+        write_rows(clickhouse_rw, dataset, rows)
         return ('ok', 200, {'Content-Type': 'text/plain'})
 
     @application.route('/tests/eventstream', methods=['POST'])
     def eventstream():
-        ensure_table_exists()
+        # TODO we need to make this work for multiple datasets
+        dataset = settings.get_dataset('events')
+        ensure_table_exists(dataset)
```

Comment:
I see this pattern in multiple places, I am concerned people may forget the second step when fetching the dataset object:

> settings.get_dataset(...)
> ensure_table_exists(dataset)

If there is no reason to instantiate a dataset with a table that does not exists, why not running the validation where we instantiate the dataset class (or call it from get_dataset)? 
If there are reasons for instantiating a dataset where the table does not exists, what about having two methods: get_enforce_dataset and get_dataset or something like that ?